Add kb articles for september. (#945)
telerik/maui-docs: knowledge-base/autocomplete-tokens-mvvm.md

---

> title: Manage Tokens Collection in .NET MAUI Autocomplete

Reviewer: ```suggestion
title: Manage the Tokens Collection in .NET MAUI AutoComplete
```

> description: Learn how to add, delete items from the AutoComplete tokens collection.

Reviewer: ```suggestion
description: Learn how to add and delete items from the AutoComplete's tokens collection.
```

type: how-to
page_title: How to Manage the Tokens collection of the RadAutoComplete for .NET MAUI in MVVM approach
slug: autocomplete-tokens-collection-mvvm
tags: autocomplete, .net maui, tokens, collection, mvvm
res_type: kb
---

## Environment

| Version | Product | Author | 
| --- | --- | ---- | 
| 7.1.0 | Telerik UI for .NET MAUI AutoComplete | [Dobrinka Yordanova](https://www.telerik.com/blogs/author/dobrinka-yordanova)| 

## Description

This article explains how to manage the tokens collection of the AutoComplete in MVVM scenario.

## Solution

The `Tokens` collection of the `AutoComplete` is a read-only collection:

```C#
public ObservableCollection<object> Tokens { get; }
```

> To add or remove tokens using MVVM setup you have to create a custom behavior.

Reviewer: ```suggestion
To add or remove tokens using an MVVM setup, create a custom behavior.
```

    {
        if (source == null || target == null)
            return;

        target.Clear();
        foreach (var o in source)
        {
            target.Add(o);
        }
    }
}
```

**2.** Set the behavior to the AutoComplete control:

```XAML
<VerticalStackLayout>
    <telerik:RadAutoComplete x:Name="autoCompleteView"
                                        ItemsSource="{Binding Source}"
                                        TextSearchPath="Name" 
                                        DisplayMode="Tokens">
        <telerik:RadAutoComplete.Behaviors>
            <local:TokensBehavior BindingContext="{Binding Source={x:Reference autoCompleteView}, Path=BindingContext}"
                                    Tokens="{Binding SelectedTokens}"/>
        </telerik:RadAutoComplete.Behaviors>
    </telerik:RadAutoComplete>
    <Button Text="Select Token" Command="{Binding SelectTokenCommand}"/>
    <Button Text="Remove Token" Command="{Binding RemoveTokenCommand}"/>
</VerticalStackLayout>
```

**3.** Add the `telerik` namespace:

```XAML
xmlns:telerik="http://schemas.telerik.com/2022/xaml/maui"
```

> **4.** Add sample model:

Reviewer: ```suggestion
**4.** Add a sample model:
```

                new Customer("Jeffery Francis", "jfrancis@mail.com"),
                new Customer("Eva Lawson", "elawson@mail.com"),
                new Customer("Emmett Santos", "esantos@mail.com"),
                new Customer("Theresa Bryan", "tbryan@mail.com"),
                new Customer("Jenny Fuller", "jfuller@mail.com"),
                new Customer("Terrell Norris", "tnorris@mail.com"),
            };

        this.SelectTokenCommand = new Command(OnSelectTokenCommandExecuted);
        this.RemoveTokenCommand = new Command(OnRemoveTokenCommandExecuted);

        this.SelectedTokens = new ObservableCollection<object>();
    }

    public ICommand SelectTokenCommand { get; set; }
    public ICommand RemoveTokenCommand { get; set; }

    public ObservableCollection<object> SelectedTokens { get; set; }

    private void OnSelectTokenCommandExecuted(object obj)
    {
        this.SelectedTokens.Add(this.Source[0]);
    }

    private void OnRemoveTokenCommandExecuted(object obj)
    {
        var item = this.SelectedTokens.FirstOrDefault();
        if (item != null)
        {
            this.SelectedTokens.Remove(item);
        }
    }
}

> ```

Reviewer: Add a See Also/Related Articles section containing links to two or three relevant articles, e.g. https://docs.telerik.com/devtools/maui/controls/autocomplete/tokens-support
